pkg/liqo-controller-manager/namespace-controller/namespace_controller.go

```diff
@@ -0,0 +1,270 @@
+	for i, namespaceMap := range namespaceMaps.Items {
+		removeMappings[namespaceMap.GetLabels()[liqo_controller_manager.VirtualNodeClusterId]] = &(namespaceMaps.Items[i])
+	}
+
+	if namespace.GetDeletionTimestamp().IsZero() {
+		if !slice.ContainsString(namespace.GetFinalizers(), namespaceFinalizer, nil) {
+			namespace.SetFinalizers(append(namespace.GetFinalizers(), namespaceFinalizer))
+			if err := r.Patch(context.TODO(), namespace, client.Merge); err != nil {
+				klog.Error(err, " --> Unable to add finalizer")
+				return ctrl.Result{}, err
+			}
+		}
+	} else {
+		if slice.ContainsString(namespace.GetFinalizers(), namespaceFinalizer, nil) {
+
+			if err := r.removeRemoteNamespaces(namespace.GetName(), removeMappings); err != nil {
+				return ctrl.Result{}, err
+			}
+
+			klog.Info(" Someone try to delete namespace, ok delete!!")
+
+			namespace.SetFinalizers(slice.RemoveString(namespace.GetFinalizers(), namespaceFinalizer, nil))
+			if err := r.Update(context.TODO(), namespace); err != nil {
+				klog.Error(err, " --> Unable to remove finalizer")
+				return ctrl.Result{}, err
+			}
+		}
+
+		klog.Info(" Namespace deleted!!")
+		return ctrl.Result{}, nil
+	}
+
+	// 1. If mapping.liqo.io label is not present there are no remote namespaces associated with this namespace, removeMappings is full
+	if remoteNamespaceName, ok := namespace.GetLabels()[mappingLabel]; ok {
+
+		// 2.a If offloading.liqo.io is present there are remote namespaces on all virtual nodes
+		if _, ok = namespace.GetLabels()[offloadingLabel]; ok {
+			klog.Infof(" Offload namespace '%s' on all remote clusters", namespace.GetName())
+			if err := r.createRemoteNamespaces(namespace, remoteNamespaceName, namespaceMaps); err != nil {
+				return ctrl.Result{}, err
+			}
+
+			for k := range removeMappings {
+				delete(removeMappings, k)
+			}
+
+		} else {
+
+			klog.Info(" Watch for virtual-nodes labels")
+
+			// 2.b Iterate on all virtual nodes labels, if the namespace has all the requested labels, is necessary to
+			// offload it onto remote cluster associated with the virtual node
+			nodes := &corev1.NodeList{}
+			if err := r.List(context.TODO(), nodes, client.MatchingLabels{liqo_controller_manager.TypeLabel: liqo_controller_manager.TypeNode}); err != nil {
+				klog.Error(err, " --> Unable to List all virtual nodes")
+				return ctrl.Result{}, client.IgnoreNotFound(err)
+			}
+
+			if len(nodes.Items) == 0 {
+				klog.Info(" No VirtualNode at the moment")
+				return ctrl.Result{}, nil
+			}
+
+			for _, node := range nodes.Items {
+				i := 0
+				dim := len(node.Labels)
+				id := node.Annotations[liqo_controller_manager.VirtualNodeClusterId]
+				for key := range node.GetLabels() {
+					i++
+					if strings.HasPrefix(key, offloadingPrefixLabel) {
+						if _, ok = namespace.GetLabels()[key]; !ok {
+							klog.Infof(" Namespace '%s' cannot be offloaded on remote cluster: %s", namespace.GetName(), id)
+							break
+						}
+					}
+
+					if i == dim {
```

Comment:
Can't you move this operation outside the inner loop and avoid using `i` and `dim`? You would increase code readability in my opinion. Do you really need them?

Reply:
I will do it, thank you